### PR TITLE
[Order] [WIP/RFC] Add state machine to order state resolver

### LINKED ISSUE #5366: [Order] StateResolver to use state machine?
The StateResolver changes the state of the order directly - this means that the state machine is not used, and no event is fired which could be used to notify something that, f.e., the payment has been completed.

Should the SM be used here? If not - what would be the best way to trigger some action when the order has been paid for?

### PR DESCRIPTION
**Pull request by dantleech**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | fixes #5366 
| License         | MIT

This PR adds the state machine to the order payment status, as decided by the StateResolver.

### CODE REVIEW

File: src/Sylius/Component/Core/OrderProcessing/StateResolver.php [hunk 1]
```diff
@@ -15,46 +15,53 @@
...
+        // do not recalculate payment state. althoguh the payment may have been
+        // refunded, the payment WAS completed for this order.
+        if ($order->getPaymentState() === OrderInterface::STATE_PAYMENT_COMPLETED) {
+            return;
+        }
+
+        // do not calculate payment state when checkout has not been completed
+        if ($order->getCheckoutState() === OrderCheckoutStates::STATE_COMPLETED) {
+            return;
+        }
```
Comment (dantleech) [inline src/Sylius/Component/Core/OrderProcessing/StateResolver.php]: The above two cases assume the COMPLETED is the final *payment state* of the order, which similifies things here, but is probably insufficent.

I guess we should take into account refunded, failed etc also.

```
if (all payments are refunded) state = REFUNDED
if (some payments are refunded) state = PARTIALLY_REFUNDED
if (some payments failed) state = PENDING
if (all paymens failed) state = FAILED
```

?

**Comment by dantleech** [comment]
Lacking tests until we agree on this or a solution.

**Comment by antonioperic** [comment]
ping @pjedrzejewski can you give some feedback on this?

**Comment by dantleech** [comment]
I see that the payment transitions already exist in `CoreBundle\..\state_machine.yml` ..

**Comment by michalmarcinkowski** [comment]
There is a pending discussion on state machine of an Order (https://github.com/Sylius/Sylius/issues/4060). Few things has changed and we will give it more thoughts soon, but any feedback is appreciated.

**Comment by dantleech** [comment]
Closing this, assuming this is going to be handled elsewhere.

**Pull request closed without merging**